File: web/src/models/tenant.ts
import { z } from "zod";

import { agentSchema, type Agent } from "@/models/agent";
import { apiKeySchema, type ApiKey } from "@/models/api-key";
import { appSchema, type App } from "@/models/app";
import { invitationSchema, type Invitation } from "@/models/invitation";
import { knowledgeSchema, type Knowledge } from "@/models/knowledge";
import { llmSchema, type Llm } from "@/models/llm";
import { memberSchema, type Member } from "@/models/member";

export interface Tenant {
  id: string;
  name: string;
  slug: string;
  icon: string | null;
  created_at: Date;
  updated_at: Date;
  members: Member[] | null;
  invitations: Invitation[] | null;
  agents: Agent[] | null;
  api_keys: ApiKey[] | null;
  clients: App[] | null;
  knowledges: Knowledge[] | null;
  llms: Llm[] | null;
}

/** Own fields only, no relations — pick/extend/partial from this in feature schemas (create/update DTOs). */
export const tenantBaseSchema = z
  .object({
    id: z.string(),
    name: z.string().min(1).max(255),
    slug: z.string().min(1).max(255),
    icon: z.string().max(255).nullable(),
    created_at: z.coerce.date(),
    updated_at: z.coerce.date(),
  })
  .strict();

export const tenantSchema: z.ZodType<Tenant> = z.lazy(() =>
  tenantBaseSchema
    .extend({
      members: z.array(memberSchema).nullable(),
      invitations: z.array(invitationSchema).nullable(),
      agents: z.array(agentSchema).nullable(),
      api_keys: z.array(apiKeySchema).nullable(),
      clients: z.array(appSchema).nullable(),
      knowledges: z.array(knowledgeSchema).nullable(),
      llms: z.array(llmSchema).nullable(),
    })
    .strict(),
);

File: internal/model/tenant.go
package model

import "time"

type Tenant struct {
	ID string `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`

	Name string  `gorm:"type:varchar(255);not null" json:"name"`
	Slug string  `gorm:"type:varchar(255);not null;unique" json:"slug"`
	Icon *string `gorm:"type:varchar(255)" json:"icon"`

	CreatedAt time.Time `gorm:"type:timestamp;not null;default:now()" json:"created_at"`
	UpdatedAt time.Time `gorm:"type:timestamp;not null;default:now()" json:"updated_at"`

	Members     []Member           `gorm:"foreignKey:TenantID;constraint:OnDelete:CASCADE" json:"members"`
	Invitations []TenantInvitation `gorm:"foreignKey:TenantID;constraint:OnDelete:CASCADE" json:"invitations"`
	Agents      []Agent            `gorm:"foreignKey:TenantID;constraint:OnDelete:CASCADE" json:"agents"`
	APIKeys     []APIKey           `gorm:"foreignKey:TenantID;constraint:OnDelete:CASCADE" json:"api_keys"`
	Clients     []App              `gorm:"foreignKey:TenantID;constraint:OnDelete:CASCADE" json:"clients"`
	Knowledges  []Knowledge        `gorm:"foreignKey:TenantID;constraint:OnDelete:CASCADE" json:"knowledges"`
	LLMs        []LLM              `gorm:"foreignKey:TenantID;constraint:OnDelete:CASCADE" json:"llms"`
}
